### PR TITLE
feat(extend/injector): bring up new extend Injector

### PR DESCRIPTION
**Pull request by SukkaW**
## What does it do?

Implementation of #4047. Should only be merged after #4048 is fixed.

## How to test

```sh
git clone -b feat-extend-inject https://github.com/sukkaw/hexo.git
cd hexo
npm install
npm test
```

## Pull request tasks

- [x] Add test cases for the changes.
- [x] Passed the CI test.

Issue resolved: #3086
Issue resolved: #4047

### CODE REVIEW

File: lib/plugins/filter/after_render/injector.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+function injectFilter(data) {
+  const { injector: _Injector } = this.extend;
+
+  function injector(data, pattern, flag, isBegin = true) {
+    return data.replace(pattern, str => {
+      if (data.includes(`hexo injector ${flag}`)) return str;
+
+      const arr = _Injector.get(flag);
+
+      if (arr.length) {
```
Comment (SukkaW) [inline lib/plugins/filter/after_render/injector.js]: If no `Injector` is registered, the `arr.length` will be `0` and nothing will be injected.

File: lib/plugins/filter/after_render/injector.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        let code = '';
+
+        for (const line of arr) {
+          code += line;
+        }
```
Comment (jiangtj) [inline lib/plugins/filter/after_render/injector.js]: How about use Array.prototype.reduce()
https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce

File: lib/extend/injector.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  get(entry) {
+    return Array.from(this.store[entry]);
+  }
+
+  register(entry, value, to = '') {
+    if (!entry) throw new TypeError('entry is required');
+    if (typeof value === 'function') value = value();
+
+    this.store[entry].add({ value, to });
```
Comment (SukkaW) [inline lib/extend/injector.js]: https://github.com/hexojs/hexo/issues/4047#issuecomment-570886958

However, `locals` is only available in `after_route_render` filter. Will be implement after #4051 is merged.

File: lib/plugins/filter/after_render/injector.js (hunk 2 above)
Comment (SukkaW) [reply lib/plugins/filter/after_render/injector.js]: https://github.com/hexojs/hexo/pull/4049/commits/fb063c58c16ab5999a64a226082a3b9bfa447050

File: lib/plugins/filter/after_render/injector.js (hunk 1 above)
Comment (jiangtj) [reply lib/plugins/filter/after_render/injector.js]: Execution `if (arr.length)` before `data.replace`, better

File: test/scripts/filters/injector.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
+'use strict';
+
+describe('Injector', () => {
+  const Hexo = require('../../../lib/hexo');
+  let hexo, injectorFilter;
+
+  beforeEach(() => {
+    hexo = new Hexo();
+    injectorFilter = require('../../../lib/plugins/filter/after_render/injector').bind(hexo);
```
Comment (SukkaW) [inline test/scripts/filters/injector.js]: I have no idea about how to invalidate the cache through each test case. And it is why the travis ci build failed...

File: lib/plugins/filter/after_render/injector.js [hunk 5]
```diff
@@ -1,53 +1,38 @@
...
+      _Injector.get(flag).filter(i => i.to === current()).map(i => i.value)
+    );
 
     if (!arr.length) return data;
 
     return data.replace(pattern, str => {
-      const code = arr.reduce((a, c) => a + c, '');
+      const code = cache.apply(
+        `${flag}-${current()}-code`,
+        arr.reduce((a, c) => a + c, '')
```
Comment (jiangtj) [inline lib/plugins/filter/after_render/injector.js]: Replace `arr.reduce((a, c) => a + c, '')` to `() => arr.reduce((a, c) => a + c, '')`, If it's not a function, it's immediately executed

Comment (SukkaW) [reply lib/plugins/filter/after_render/injector.js]: Rebased and updated.

**Comment by SukkaW** [comment]
@hexojs/core It is ready to be reviewed now.

File: lib/extend/injector.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  get(entry) {
+    return Array.from(this.store[entry]);
+  }
+
+  register(entry, value, to = 'default') {
+    if (!entry) throw new TypeError('entry is required');
+    if (typeof value === 'function') value = value();
+
+    this.store[entry].add({ value, to });
```
[the remark this answers is not in the document]
Comment (dailyrandomphoto) [reply lib/extend/injector.js]: To improve performance:
```js
...
    this.store = {
      head_begin: {},
...

register(entry, value, to = 'default') {
    ...
    const entryMap = this.store[entry] || this.store['body_end'];
    const arr = entryMap[to] || []; // or new Set();
    arr.push(value);
    entryMap[to] = arr;
```
```js
  get(entry, to = 'default') {
    const entryMap = this.store[entry] || this.store['body_end'];
    return entryMap[to];
  }

  getText(entry, to = 'default') {
    const arr = this.get(entry, to);
    if (!arr || !arr.length) return '';
    return arr.join('');
  }
```
then `cache.apply` in `lib/plugins/filter/after_render/injector.js` is not required.

File: lib/plugins/filter/after_render/injector.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      `${flag}-${current()}-arr`,
+      _Injector.get(flag).filter(i => i.to === current()).map(i => i.value)
+    );
+
+    if (!arr.length) return data;
+
+    return data.replace(pattern, str => {
+      const code = cache.apply(
+        `${flag}-${current()}-code`,
+        () => arr.reduce((a, c) => a + c, '')
```
Comment (dailyrandomphoto) [inline lib/plugins/filter/after_render/injector.js]:
```js
arr.reduce((a, c) => a + c, '')
=>
arr.join('');
```

File: lib/plugins/filter/after_render/injector.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    if (page.tag) return 'tag';
+    return 'default';
+  };
+
+  function injector(data, pattern, flag, isBegin = true) {
+    if (data.includes(`hexo injector ${flag}`)) return data;
+
+    const arr = cache.apply(
+      `${flag}-${current()}-arr`,
+      _Injector.get(flag).filter(i => i.to === current()).map(i => i.value)
```
Comment (dailyrandomphoto) [inline lib/plugins/filter/after_render/injector.js]: The size of `arr` is not large, the use of cache may not have much improvement.

File: lib/plugins/filter/after_render/injector.js (hunk 7 above)
Comment (dailyrandomphoto) [reply lib/plugins/filter/after_render/injector.js]: The size of `arr` is not large, the use of cache may not have much improvement.

File: lib/plugins/filter/after_render/injector.js (hunk 8 above)
Comment (SukkaW) [reply lib/plugins/filter/after_render/injector.js]: After enabling cache, benchmark (performed by travis ci) shows the time drops from `16.3s` to `15.8s` on average.

It is not about the size of array, it is about how many times the `injector()` will be executed.

File: lib/plugins/filter/after_render/injector.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  const current = () => {
+    const { page } = locals;
+
+    if (page.__index) return 'home';
+    if (page.__post) return 'post';
+    if (page.__page) return 'page';
+    if (page.archive) return 'archive';
+    if (page.category) return 'category';
+    if (page.tag) return 'tag';
+    return 'default';
```
Comment (dailyrandomphoto) [inline lib/plugins/filter/after_render/injector.js]: How about support layout(`page.layout`) as page type.
Some themes have `tags`, `categories`, `search` layouts.
``` diff
    if (page.tag) return 'tag';
+   if (page.layout) return page.layout;
```

File: lib/extend/injector.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    if (!arr || !arr.length) return '';
+    return arr.join('');
+  }
+
+  register(entry, value, to = 'default') {
+    if (!entry) throw new TypeError('entry is required');
+    if (typeof value === 'function') value = value();
+
+    const entryMap = this.store[entry] || this.store.head_end;
+    const valueSet = entryMap[to] || new Set();
```
Comment (jiangtj) [inline lib/extend/injector.js]: 似乎因为我的这个评论 https://github.com/hexojs/hexo/issues/4047#issuecomment-570886958 改了好多 😂 

这里的map是为了更好的性能，我这样理解没错吧，但是在我的观点中，性能与维护成本，拓展性之类的需要一起考虑

to参数对page类型的封装，好处是仅暴露这个风险小，同时map的引入，也提高了性能（我的观点和下面一样，由于arr不大，改善应该也不大），但相对的拓展性也减少了

正好有个现成的例子（最近打算些）

我想实现一个mermaid的插件，拓展marked的code渲染函数实现，由于mermaid不支持node端渲染，需要同时在浏览器中引入（注入）js，但这个js脚本很大，大概1mb，我打算在front-matter中设置变量来决定其是否使用，那么这时候又该怎么利用这个api呢

在我评论的方案中只需要 （locals）=> locals.page.mermaid即可

https://github.com/mermaid-js/mermaid

pad上写的，翻译有点麻烦，暂时中文，明天再翻。。。

Comment (SukkaW) [reply lib/extend/injector.js]: `Injector` is designed to fix #3086. `Injector` should only be used to insert static HTML snippet. Any features should be implemented using theme if can.

**Comment by dailyrandomphoto** [comment]
LGTM!

**Comment by SukkaW** [comment]
@dailyrandomphoto @jiangtj 

I have refactored the `Injector` recently.

A new `Injector.exec()` method has been added (`libs/plugins` dir should only be used for internal plugins that are based on `hexo.extend`).
Also, `Injector` is now separate from `Filter`. `Injector` will insert code before `after_router_render` filter's execution.

The test on my local machine with a dummy hexo site looks good as well:

image

**Comment by SukkaW** [comment]
Rebased and ready for review.

**Comment by SukkaW** [comment]
Resolved conflict & ready for review.

**Pull request merged**